File: 146-lru-cache/146-lru-cache.java
class LRUCache {
    class DLLNode {
        int key, val;
        DLLNode prev, next;
        public DLLNode(){};
        public DLLNode(int key, int val) {
            this.key = key;
            this.val = val;
        }
    }
    
    int size;
    int capacity;
    DLLNode head, tail;
    Map<Integer, DLLNode> cache;
    
    public LRUCache(int capacity) {
        this.size = 0;
        this.capacity = capacity;
        head = new DLLNode();
        tail = new DLLNode();
        head.next = tail;
        tail.prev = head;
        cache = new HashMap<>();
    }
    
    public int get(int key) {
        if(cache.containsKey(key)) {
            DLLNode node = cache.get(key);
            moveToHead(node);
            return node.val;
        }
        return -1;
    }
    public void moveToHead(DLLNode node) {
        removeNode(node);
        addToHead(node);
    }
    public void removeNode(DLLNode node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }
    
    public void put(int key, int value) {
        if(cache.containsKey(key)) {
            DLLNode node = cache.get(key);
            node.val = value;
            moveToHead(node);
            return;
        } else {
            DLLNode node = new DLLNode(key, value);
            cache.put(key, node);
            addToHead(node);
            size++;
            if(size > capacity) {
                DLLNode lastNode = removeTail();
                cache.remove(lastNode.key);
                size--;
            }
        }
    }
    public DLLNode removeTail() {
        DLLNode node = tail.prev;
        removeNode(node);
        return node;
    }
    public void addToHead(DLLNode node) {
        node.next = head.next;
        node.prev = head;
        node.next.prev = node;
        head.next = node;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 146-lru-cache/146-lru-cache.py
class Node:
    def __init__(self, key = 0, val = 0, next = None, prev = None):
        self.key = key
        self.val = val
        self.next = next
        self.prev = prev
        
class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.size = 0
        self.cache = {}
        self.head, self.tail = Node(), Node()
        self.head.next = self.tail
        self.tail.prev = self.head
        
    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        else:
            node = self.cache[key]
            self.moveToHead(node)
        return node.val
    
    def moveToHead(self,node):
        self.deleteNode(node)
        self.addToHead(node)
    
    def deleteNode(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev
    
    def addToHead(self, node):
        node.next = self.head.next
        node.prev = self.head
        
        self.head.next.prev = node
        self.head.next = node

    def removeTail(self):
        tail = self.tail.prev
        self.deleteNode(tail)
        return tail
        
    def put(self, key: int, value: int) -> None:
        node = self.cache.get(key)
        if not node:
            node = Node(key, value)
            self.cache[key] = node
            
            self.addToHead(node)
            self.size += 1
            if self.size > self.capacity:
                tail= self.removeTail()
                self.cache.pop(tail.key)
                self.size -= 1
        else:
            node.val = value
            self.moveToHead(node)
            
        
        
            


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)